### PR TITLE
[Serde reflection] Add visitors + docs

### PR DESCRIPTION
**Pull request by matbd**
## Motivation

* For code generation, I need generic visitors. This adds them and refactor `normalize`.
* I realized we cannot easily access the fully qualified name of types, so scrape the TODO.

## Testing

```
cargo x test -p serde-reflection
```

### CODE REVIEW

**Comment by bmwill** [comment]
Hmm Yeah even `https://doc.rust-lang.org/stable/std/any/struct.TypeId.html` has the limitation that the type must be `'static` which wont work for borrowed types.

File: common/serde-reflection/src/format.rs [hunk 1]
```diff
@@ -167,45 +241,75 @@ impl FormatHolder for VariantFormat {
...
-            }
-        }
-    }
 }
 
 impl<T> FormatHolder for Named<T>
 where
     T: FormatHolder + std::fmt::Debug,
 {
+    fn visit(&self, f: &mut dyn FnMut(&Format) -> Result<()>) -> Result<()> {
```
Comment (bmwill) [inline common/serde-reflection/src/format.rs]: Having the closure being taken in as an `&mut` was a little weird to see but I do understand why you're doing it that way.

File: common/serde-reflection/src/format.rs [hunk 2]
```diff
@@ -103,7 +103,7 @@ pub enum VariantFormat {
 
 pub trait FormatHolder {
     /// Visit all the formats in `self` in a depth-first way.
-    fn visit(&self, f: &mut dyn FnMut(&Format) -> Result<()>) -> Result<()>;
+    fn visit<'a>(&'a self, f: &mut dyn FnMut(&'a Format) -> Result<()>) -> Result<()>;
```
Comment (bmwill) [inline common/serde-reflection/src/format.rs:106]: What's the reason why these lifetime annotations are needed?

Comment (matbd) [reply common/serde-reflection/src/format.rs:106]: Fair question. I updated the last commit with a better unit test to show the difference.

**Pull request closed without merging**